[LSR] Don't count conditional loads/store as enabling pre/post-index

john-brawn-arm opened the pull request:
When a load/store is conditionally executed in a loop it isn't a candidate for pre/post-index addressing, as the increment of the address would only happen on those loop iterations where the load/store is executed.

Detect this and only discount the AddRec cost when the load/store is unconditional.

john-brawn-arm committed:
[LSR] Don't count conditional loads/store as enabling pre/post-index

When a load/store is conditionally executed in a loop it isn't a
candidate for pre/post-index addressing, as the increment of the
address would only happen on those loop iterations where the
load/store is executed.

Detect this and only discount the AddRec cost when the load/store is
unconditional.

john-brawn-arm committed:
clang-format

artagnon reviewed (commented):
Makes sense to me.

nikic commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1647,6 +1659,12 @@ bool LSRFixup::isUseFullyOutsideLoop(const Loop *L) const {
+/// Test whether this fixup is for an instruction that's unconditional, i.e.
+/// it's executed in every loop iteration.
+bool LSRFixup::isUseUnconditional(const Loop *L) const {
+  return isGuaranteedToExecuteForEveryIteration(UserInst, L);
```
Hm... what kind of "unconditional" execution do you actually need in this context? There's a hierarchy here:

1. Not inside interior control flow in the loop (your test case).
2. Plus no early exits.
3. Plus no abnormal exits.

I think that at least you don't care about abnormal exits in this context? If there is an instruction that can throw/diverge in the loop, that shouldn't interfere with indexed addressing, right? In that case using isGuaranteedToExecuteForEveryIteration() is both unnecessarily conservative and unnecessarily expensive, as it has to scan for abnormal exits.

john-brawn-arm replied:
What we ultimately want to check for is situations where, when we reach codegen, the load/store and the add will be combined into a single pre/post increment load/store. For this to happen during ISel the load/store and the add need to be in the same block. Looking into this a bit, LSRInstance::ImplementSolution is where we decide the InsertPos, but this currently only allows the loop header and the IVIncInsertPos, which will be the loop latch in most cases.

I think that it should be possible to change this to insert the add instruction in the LSRUse block if it dominates IVIncInsertPos, as when that's the case we must go through the LSRUse block to reach IVIncInsertPos, so an instruction inserted there is guaranteed to be available at IVIncInsertPos. I've changed this patch to check for that, and I'll be doing a follow-on patch to change LSRInstance::ImplementSolution to hoist to this location.

john-brawn-arm committed:
Use m_scev_APInt

john-brawn-arm committed:
Update minloop.ll test

john-brawn-arm committed:
Alternate approach to determining if fixup is always executed

Instead of isGuaranteedToExecuteForEveryIteration, check that the block the
instruction is in dominates the block the IV increment is in.

john-brawn-arm commented:
Ping.

nikic commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1318,6 +1318,11 @@ class LSRUse {
 
+  /// This records whether all of the fixups using this LSRUse are unconditional
+  /// within the loop, meaning they will be executed in every iteration of the
+  /// loop.
```
I'd like some more precision here, something along these lines:
```suggestion
  /// within the loop, meaning they will be executed on every path to the loop latch.
  /// This includes fixups before early exits.
```

nikic reviewed (approved):
LGTM

john-brawn-arm committed:
Adjust comment.
